File: src/main/java/oscar/cp/constraints/Or.java
/*******************************************************************************
 * OscaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *   
 * OscaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License  for more details.
 *   
 * You should have received a copy of the GNU Lesser General Public License along with OscaR.
 * If not, see http://www.gnu.org/licenses/lgpl-3.0.en.html
 ******************************************************************************/
package oscar.cp.constraints;

import oscar.cp.core.CPOutcome;
import oscar.cp.core.CPPropagStrength;
import oscar.cp.core.CPVarBool;
import oscar.cp.core.CPVarInt;
import oscar.cp.core.Constraint;
import oscar.reversible.ReversibleBool;
import oscar.reversible.ReversibleInt;

/**
 * Logical Or Constraint
 * @author Pierre Schaus pschaus@gmail.com
 */
public class Or extends Constraint {
	
	private CPVarBool [] x;
	private CPVarBool y;
	
	ReversibleInt nbBound;
	ReversibleBool ytrue;
	


    /**
     * y is true if at least one of the xi's is true, false otherwise
     * @param x
     * @param y
     */
	public Or(CPVarBool [] x, CPVarBool y) {
		super(x[0].s());
		this.x = x;
		this.y = y;
		nbBound = new ReversibleInt(s,0); // number of values assigned to false
		ytrue = new ReversibleBool(s);
		ytrue.setValue(false);
	}

	@Override
	protected CPOutcome setup(CPPropagStrength l) {
		for (int i = 0; i < x.length; i++) {
			if (x[i].isTrue()) {
				if (y.assign(1) == CPOutcome.Failure) {
					return CPOutcome.Failure;
				} else {
					return CPOutcome.Success;
				}
			}
		}
		// we know no variables from X are bound to 1
		for (int i = 0; i < x.length; i++) {
			if (!x[i].isBound()) {
				x[i].callValBindIdxWhenBind(this, i);
			} else {
				assert(x[i].isFalse());
				nbBound.incr();
			}
		}
		if (!y.isBound()) {
			if (nbBound.getValue() == x.length) {
				if (y.assign(0) == CPOutcome.Failure) {
					return CPOutcome.Failure;
				}
			}
			y.callValBindWhenBind(this);
		}
		else {
			if (y.getValue() == 0) {
				for (int i = 0; i < x.length; i++) {
						if (x[i].assign(0) == CPOutcome.Failure) {
							return CPOutcome.Failure;
						}
				}
				return CPOutcome.Success;
			} else {
				ytrue.setValue(true);
				if (nbBound.getValue() == x.length-1) {
					for (int i = 0; i < x.length; i++) {
						if (!x[i].isBound()) {
							if (x[i].assign(1) == CPOutcome.Failure) {
								return CPOutcome.Failure;
							}
							return CPOutcome.Success;
						}
					}
				}
			}
		}
		return CPOutcome.Suspend;
	}
	
	
	@Override
	protected CPOutcome valBindIdx(CPVarInt var, int idx) {
		if (var.getValue() == 1) {
			if (y.assign(1) == CPOutcome.Failure) {
				return CPOutcome.Failure;
			}
			return CPOutcome.Success;
		} else {
			nbBound.incr();			
			if (nbBound.getValue() == x.length) {
				if (y.assign(0) == CPOutcome.Failure) {
					return CPOutcome.Failure;
				}
			} else if (nbBound.getValue() == x.length-1 && ytrue.getValue()){
				for (int i = 0; i < x.length; i++) {
					if (!x[i].isBound()) {
						if (x[i].assign(1) == CPOutcome.Failure) {
							return CPOutcome.Failure;
						}
						return CPOutcome.Success;
					}
				}
				return CPOutcome.Suspend;
			}
		}
		return CPOutcome.Suspend;
	}
	
	@Override
	protected CPOutcome valBind(CPVarInt yvar) {
		if (yvar.getValue() == 0) {
			for (int i = 0; i < x.length; i++) {
					if (x[i].assign(0) == CPOutcome.Failure) {
						return CPOutcome.Failure;
					}
				
			}
			return CPOutcome.Success;
		} else {
			ytrue.setValue(true);
			if (nbBound.getValue() == x.length-1){
				for (int i = 0; i < x.length; i++) {
					if (!x[i].isBound()) {
						if (x[i].assign(1) == CPOutcome.Failure) {
							return CPOutcome.Failure;
						}
						return CPOutcome.Success;
					}
				}
			}
		}
		return CPOutcome.Suspend;
	}

}

File: oscar-cp/src/main/scala/oscar/cp/constraints/Or.scala
/*******************************************************************************
 * OscaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *   
 * OscaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License  for more details.
 *   
 * You should have received a copy of the GNU Lesser General Public License along with OscaR.
 * If not, see http://www.gnu.org/licenses/lgpl-3.0.en.html
 ******************************************************************************/
package oscar.cp.constraints

import oscar.cp.core._
import oscar.algo.reversible._
import oscar.cp.core.CPOutcome._
import oscar.algo.reversible.ReversibleSparseSet

/**
 * or x_i = true
 * @author Pierre Schaus pschaus@gmail.com
 */
class Or(val x: Array[CPBoolVar]) extends Constraint(x(0).store, "Or") {

  var i = 0
  var down = 0
  var up = 0
  val n = x.size

  override def setup(l: CPPropagStrength): CPOutcome = {
    i = 0
    while (i < n && x(i).isBound) {
      if (x(i).isBoundTo(1)) return Success
      i += 1
    }
    down = i
    i = n-1
    while (i >= 0 && x(i).isBound && i >= down) {
      if (x(i).isBoundTo(1)) return Success
      i -= 1
    }
    up = i
    if (down > up) {
      return Failure
    } else if (down == up) {
       x(up).assign(1)
      return Success
    } else {
      assert(down != up)
      x(down).callValBindIdxWhenBind(this,down)
      x(up).callValBindIdxWhenBind(this,up)
      
      //x(up).callPropagateWhenBind(this,false)
      //x(down).callPropagateWhenBind(this,false)
    }
    Suspend
  }

  override def valBindIdx(y: CPIntVar, idx: Int): CPOutcome = {
    if (y.isBoundTo(1)) return Success
    if (down >= n || up < 0 || x(down).isBound || x(up).isBound || (down >= up)) {
      i = 0
      while (i < n && x(i).isBound) {
        if (x(i).isBoundTo(1)) return Success
        i += 1
      }
      down = i
      i = n - 1
      while (i >= 0 && x(i).isBound && i >= down) {
        if (x(i).isBoundTo(1)) return Success
        i -= 1
      }
      up = i
    } 
    if (down > up) {
      return Failure
    } else if (down == up) { // only one unassigned var
      x(up).assign(1)
      return Success
    } else {
      assert(down != up)
      assert(x(down).isBound == false)
      assert(x(up).isBound == false)
      x(down).callValBindIdxWhenBind(this,down)
      x(up).callValBindIdxWhenBind(this,up)
    }
    Suspend
  }
  
  override def propagate(): CPOutcome = {
    if (down < n && x(down).isBoundTo(1)) return Success
    if (up >= 0 && x(up).isBoundTo(1)) return Success
    if (down >= n || up < 0 || x(down).isBound || x(up).isBound || (down >= up)) {
      i = 0
      while (i < n && x(i).isBound) {
        if (x(i).isBoundTo(1)) return Success
        i += 1
      }
      down = i
      i = n - 1
      while (i >= 0 && x(i).isBound && i >= down) {
        if (x(i).isBoundTo(1)) return Success
        i -= 1
      }
      up = i
    } 
    if (down > up) {
      return Failure
    } else if (down == up) { // only one unassigned var
      x(up).assign(1)
      return Success
    } else {
      assert(down != up)
      assert(x(down).isBound == false)
      assert(x(up).isBound == false)
      x(down).callPropagateWhenBind(this,false)
      x(up).callPropagateWhenBind(this,false)
    }
    Suspend
  }  
}

